src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.cs

```diff
@@ -165,5 +166,169 @@ public sealed override byte[] GetSubjectKeyIdentifier(X509Certificate2 certifica
                 return ski;
             }
         }
+
+        public override T GetPrivateKeyForSigning<T>(X509Certificate2 certificate, bool silent)
+        {
+            return GetPrivateKey<T>(certificate, silent, preferNCrypt: true);
+        }
+
+        public override T GetPrivateKeyForDecryption<T>(X509Certificate2 certificate, bool silent)
+        {
+            return GetPrivateKey<T>(certificate, silent, preferNCrypt: false);
+        }
+
+        private T GetPrivateKey<T>(X509Certificate2 certificate, bool silent, bool preferNCrypt) where T : AsymmetricAlgorithm
+        {
+            if (!certificate.HasPrivateKey)
+            {
+                return null;
+            }
+
+            SafeProvOrNCryptKeyHandle handle = GetCertificatePrivateKey(
+                certificate,
+                silent,
+                preferNCrypt,
+                out CryptKeySpec keySpec,
+                out Exception exception);
+
+            using (handle)
```

Comment:
Question: You can `using (null)`? Interesting.

Reply:
Yep, the finally block `using` generates is null-safe.